crates/bevy_asset/src/server/loaders.rs

```diff
@@ -0,0 +1,152 @@
+                    Some(receiver.recv().await.unwrap())
+                }
+            }
+        }
+    }
+
+    /// Get an [`AssetLoader`] by its [`std::any::type_name`]. This resolves to `None` immediately if
+    /// no loader has been registered by that name.
+    ///
+    /// If a loader has been pre-registered for this name, then it will resolve once it is available.
+    pub fn get_by_loader_type_name(
+        &self,
+        name: &str,
+    ) -> impl Future<Output = Option<AssetLoaderSmartPointer>> + 'static {
+        let type_id = self.type_name_to_type_id.get(name).copied();
+        let loader = type_id.map(|type_id| self.get_by_asset_type_id(type_id));
+
+        async { loader?.await }
+    }
+
+    /// Get an [`AssetLoader`] by a file extension it supports. This resolves to `None` immediately if
+    /// no loader has been registered for that extension.
+    ///
+    /// If a loader has been pre-registered for this extension, then it will resolve once it is available.
+    pub fn get_by_extension(
+        &self,
+        extension: &str,
+    ) -> impl Future<Output = Option<AssetLoaderSmartPointer>> + 'static {
+        let type_id = self.extension_to_type_id.get(extension).copied();
+        let loader = type_id.map(|type_id| self.get_by_asset_type_id(type_id));
+
+        async { loader?.await }
+    }
+
+    /// Registers a new [`AssetLoader`]. [`AssetLoader`]s must be registered before they can be used.
+    pub fn register_loader<L: AssetLoader>(&mut self, loader: L) {
+        let type_name = std::any::type_name::<L>();
+        let loader = Arc::new(loader);
+
+        let (type_id, is_new) = if let Some(index) = self.preregistered_loaders.remove(type_name) {
+            (index, false)
+        } else {
+            (TypeId::of::<L::Asset>(), true)
+        };
+
+        for extension in loader.extensions() {
+            self.extension_to_type_id
+                .insert(extension.to_string(), type_id);
+        }
+
+        if is_new {
+            self.type_name_to_type_id.insert(type_name, type_id);
+            self.values.insert(type_id, MaybeAssetLoader::Ready(loader));
+        } else {
+            let maybe_loader = std::mem::replace(
+                self.values.get_mut(&type_id).unwrap(),
+                MaybeAssetLoader::Ready(loader.clone()),
+            );
+            match maybe_loader {
+                MaybeAssetLoader::Ready(_) => unreachable!(),
+                MaybeAssetLoader::Pending { sender, .. } => {
+                    IoTaskPool::get()
+                        .spawn(async move {
+                            let _ = sender.broadcast(loader).await;
+                        })
+                        .detach();
+                }
+            }
+        }
+    }
+
+    /// Pre-register a loader that will later be added.
+    ///
+    /// Assets loaded with matching extensions will be blocked until the
+    /// real loader is added.
```

Comment:
Could we hint users when pre-registration is necessary?

I remember at one point I stumbled upon a situation where I needed to register a loader at a specific part of Bevy's app init phase, and it took a while to figure out, so any hints are helpful (especially if it's related to that).

Reply:
I'm not too sure how to go about this, so I'm going to palm this off onto a future PR for now.